lib/afinter.c

```diff
@@ -124,14 +125,70 @@ afinter_source_post(gpointer s)
   afinter_source_update_watches(self);
 }
 
+static void afinter_source_start_watches(AFInterSource *self);
+static void afinter_source_stop_watches(AFInterSource *self);
+
+static void
+afinter_source_run(gpointer s)
+{
+  AFInterSource *self = (AFInterSource *) s;
+
+  iv_init();
+
+  /* post event is used by other threads and can only be unregistered if
+   * current_afinter_source is set to NULL in a thread safe manner */
+  iv_event_register(&self->post);
+  iv_event_register(&self->schedule_wakeup);
+  iv_event_register(&self->exit);
+
+  g_static_mutex_lock(&internal_msg_lock);
+  self->free_to_send = TRUE;
+  g_static_mutex_unlock(&internal_msg_lock);
+
+  afinter_source_start_watches(self);
+  afinter_source_update_watches(self);
+
+  iv_main();
+
+  iv_event_unregister(&self->exit);
+  iv_event_unregister(&self->post);
```

Comment:
I have a crash because of a possible race, where an internal message triggers the post event, but the it has previously been unregistered.

I'm investigating this scenario.